File: backend/app/models/flashcard.py
from sqlalchemy import Column, Integer, String, Text, DateTime, ForeignKey, Enum
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from ..database import Base
import enum

class DifficultyLevel(str, enum.Enum):
    EASY = "easy"
    MEDIUM = "medium"
    HARD = "hard"

class Flashcard(Base):
    __tablename__ = "flashcards"

    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id", ondelete="CASCADE"), nullable=False)
    question = Column(Text, nullable=False)
    answer = Column(Text, nullable=False)
    subject = Column(String(100), nullable=True, index=True)
    difficulty = Column(Enum(DifficultyLevel), default=DifficultyLevel.MEDIUM)
    times_reviewed = Column(Integer, default=0)
    times_correct = Column(Integer, default=0)
    next_review = Column(DateTime(timezone=True), nullable=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())

    #relationship
    user = relationship("User", back_populates="flashcards")

File: frontend/lib/models/flashcard.dart
class Flashcard {
  final int id;
  final int userId;
  final String question;
  final String answer;
  final String? subject;
  final String difficulty;
  final int timesReviewed;
  final int timesCorrect;
  final DateTime? nextReview;
  final DateTime createdAt;

  Flashcard({
    required this.id,
    required this.userId,
    required this.question,
    required this.answer,
    this.subject,
    required this.difficulty,
    required this.timesReviewed,
    required this.timesCorrect,
    this.nextReview,
    required this.createdAt,
  });

  factory Flashcard.fromJson(Map<String, dynamic> json) {
    return Flashcard(
      id: json['id'],
      userId: json['user_id'],
      question: json['question'],
      answer: json['answer'],
      subject: json['subject'],
      difficulty: json['difficulty'],
      timesReviewed: json['times_reviewed'],
      timesCorrect: json['times_correct'],
      nextReview: json['next_review'] != null
          ? DateTime.parse(json['next_review'])
          : null,
      createdAt: DateTime.parse(json['created_at']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'question': question,
      'answer': answer,
      'subject': subject,
      'difficulty': difficulty,
    };
  }

  double get accuracy {
    if (timesReviewed == 0) return 0;
    return (timesCorrect / timesReviewed) * 100;
  }
}
